### PR TITLE
#163383187 Add functionality to favourite and unfavourite an article 

### PR DESCRIPTION
**Pull request by oma0256**
**What does this PR do?**
 - Allow an authenticated user to favourite and unfavourite an article.
 - Prevent an unauthenticated user from favouriting and unfavouriting an article.

**How should this be manually tested?**
 - Download and install postman onto you're local machine
 - Clone the repo onto you're local machine
 - Create and activate virtual environment
 - Move into the project directory
 `
pip3 install -r requirements.txt
python3 manage.py runserver
`
- Open postman
- Register a user using the route 
`http://127.0.0.1:8000/api/v1/users/`
- Login the user you just registered
`http://127.0.0.1:8000/api/v1/users/login/`
- Get the token that was returned
- Insert a header in postman with a key of `Authorization`  and value `Bearer <token>`
- Then visit the route `http://127.0.0.1:8000/api/v1/articles` to create an article using `POST`.
- Then visit the route `http://127.0.0.1:8000/api/v1/articles/<slug>/favorite` to either favourite or unfavourite an article `GET`.

**What are the relevant pivotal tracker stories?**
[#163383187](https://www.pivotaltracker.com/n/projects/2240000/stories/163383187)

### CODE REVIEW

**Review by fahdjamy — CHANGES_REQUESTED** [review]
This is some pretty work @oma0256, but if you could add more comments to your methods and classes to make your methods and classes easier to understand by a person reading through your code a great suggestion that was also given to me by @malep2007 on my create-article PR.
Others this is good work.

File: authors/apps/articles/urls.py [hunk 1]
```diff
@@ -9,4 +9,6 @@
          name='articles'),
     path('<slug>', views.ArticleDetailApiView.as_view(),
          name='article-details'),
+    path('<slug>/favorite', views.ToggleFavoriteAPIView.as_view(),
```
Comment (missvicki) [inline authors/apps/articles/urls.py:24]: @oma0256 we have been following the naming convention on the urls with the trailing slash. Could you please revise this

File: authors/apps/articles/views.py [hunk 2]
```diff
@@ -88,3 +94,28 @@ def delete(self, request, slug):
...
+
+class ToggleFavoriteAPIView(generics.GenericAPIView):
+    """
+    ToggleFavoriteAPIView favorites and unfavorites an article
+    """
+    permission_classes = (permissions.IsAuthenticated,)
+    serializer_class = serializers.ArticleSerializer
+    renderer_classes = (ArticleJSONRenderer,)
+
+    def get_object(self, slug):
```
Comment (missvicki) [inline authors/apps/articles/views.py]: this is going to bring about duplication, @FahdJamy defined the same function in one of his views. Please revise.

**Review by missvicki — CHANGES_REQUESTED** [review]

**Comment by Karuhanga** [comment]
Good work there @oma0256.
I'd just like to make two observations;
- I think you might have forgotten to `rebase` after the lastest merge.
- The new migrations seem to have replaced the previous ones, which might cause breaking changes. Isn't there a better way of doing this?

**Review by karuhanga — CHANGES_REQUESTED** [review]
Good work there @oma0256.
I'd just like to make two observations;
- I think you might have forgotten to `rebase` after the lastest merge.
- The new migrations seem to have replaced the previous ones, which might cause breaking changes. Isn't there a better way of doing this?

File: authors/apps/articles/urls.py (hunk 1 above)
Comment (oma0256) [reply authors/apps/articles/urls.py:24]: @missvicki we're actually currently not using the trailing slash on our urls

**Review by karuhanga — APPROVED** [review]

**Review by fahdjamy — APPROVED** [review]
thanks for implementing the changes @oma0256.
This LGTM now.

**Comment by malep2007** [comment]
There are merge conflicts, in this PR, can this be sorted @oma0256

**Comment by oma0256** [comment]
@malep2007 Let me get right on it

**Comment by oma0256** [comment]
@malep2007 fixed the conflicts

File: authors/apps/articles/urls.py (hunk 1 above)
Comment (karuhanga) [reply authors/apps/articles/urls.py:24]: True @oma0256 .

**Review by missvicki — APPROVED** [review]

**Pull request merged**